File: docs/DSA/src/py/median_of_two_sorted_arrays.py
class Solution:
    def findMedianSortedArrays(self, nums1: list[int], nums2: list[int]) -> float:
        A, B = nums1, nums2
        if len(A) > len(B):
            A, B = B, A

        half = (len(A)+len(B))//2
        left, right = 0, len(A)-1
        while True:
            midA = left + (right-left)//2
            midB = half - midA - 2
            Aleft = A[midA] if midA >= 0 else float('-inf')
            Aright = A[midA+1] if midA+1 < len(A) else float('inf')
            Bleft = B[midB] if midB >= 0 else float('-inf')
            Bright = B[midB+1] if midB+1 < len(B) else float('inf')
            if Aleft <= Bright and Bleft <= Aright:
                if (len(A)+len(B))%2 == 0:
                    return (max(Aleft,Bleft) + min(Aright, Bright))/2
                else:
                    return min(Aright, Bright)
            elif Aleft > Bright:
                right = midA-1
            else:
                left = midA+1
        

File: docs/DSA/src/go/median_of_two_sorted_arrays.go
package main

import "math"

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	if len(nums1) > len(nums2) {
		nums1, nums2 = nums2, nums1
	}

	m, n := len(nums1), len(nums2)
	half := (m + n + 1) / 2

	low, high := 0, m

	for low <= high {
		partA := (low + high) / 2
		partB := half - partA

		var leftA, rightA int
		if partA == 0 {
			leftA = math.MinInt64
		} else {
			leftA = nums1[partA-1]
		}
		if partA == m {
			rightA = math.MaxInt64
		} else {
			rightA = nums1[partA]
		}

		var leftB, rightB int
		if partB == 0 {
			leftB = math.MinInt64
		} else {
			leftB = nums2[partB-1]
		}
		if partB == n {
			rightB = math.MaxInt64
		} else {
			rightB = nums2[partB]
		}

		if leftA <= rightB && leftB <= rightA {
			if (m+n)%2 == 0 {
				return (float64(max(leftA, leftB)) + float64(min(rightA, rightB))) / 2
			}
			return float64(max(leftA, leftB))
		} else if leftA > rightB {
			high = partA - 1
		} else {
			low = partA + 1
		}
	}
	return 0
}
